39139 larger gsasii refactor

JackEAllen opened the pull request:
### Description of work

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
- Try to encapsulate and centralise logic for managing GSAS-II input parameters, paths, and configurations into a class (`GSAS2Handler`).
- Encapsulate related attributes into logical groups using dataclasses to improve readability and reduce the size of the inits for `GSAS2Handler` and `GSAS2Model`.
- Add additional tests to check GSAS2 is correctly loaded and necessary parameters are handles correctly before GSAS2 is imported. 
- Improve modularity and maintainability of model through type driven refactoring - (this is not an extensive refactor as this would be out of scope. Additional refactoring can be done in the future to further improve modularity and maintainability).

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->
Older versions of GSAS2 are no longer available meaning newer versions will be more likely used within Mantid going forward for users who do not use IDAaaS. In newer version of GSAS2, the `GSASIIscriptable.py` was relocated which caused a failure to import GSAS2 depending on the version of GSAS2 installed. Dynamically finding `GSASIIscriptable` has since been resolved as part of #38882 which will cover ONLY changes to the location of `GSASIIscriptable`. 

In case the location of any other necessary files required to import GSAS2 occurs going forward, centralising how GSAS2 is configured is required to improve the speed at which we can resolve such changes going forward, taking a more dynamic approach by tweaking `limited_rglob()` or related method instead of relying on hard-coded paths. This PR achieves this and aims to provide a little more flexibility for how these files are found by utilising `limited_rglob()` to find all files required to import GSAS2. 

In the spirit of refactoring as and when code is touched, I have chosen the add types to all attributes within the model and performed type driven refactoring as part of this PR. Note that further refactoring could be done, but should be completed as part of a separate issue to avoid this PR growing too big.

Fixes #39139  . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
[Mantid anaconda label](https://anaconda.org/mantid/mantidworkbench): `new_gsas_path_search_test` - **PLEASE REMOVE LABEL FROM ANACONDA ONCE PR IS APPROVED.**

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->
#### Newer Versions of GSASII (`5758` and later)
1.  Follow [test 11](https://developer.mantidproject.org/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.html#id15) from the [Engineering Diffraction Testing Instructions] and verify:(https://developer.mantidproject.org/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.html#engineering-diffraction-testing)
- [x] A standard refinement works]
- [x] Moving the` x_min` and `x_max` limits to `18401.18` and `44658.51` respectively and performing a refinement only fits to the newly defined range.
- [x] Setting `Override Unit Cell Length` to `3.65` and clicking refine works as expected.
- [x] Performing a refinement with `Microstrain`, `Sigma-1` and `Gamma (Y)` selected works as expected.

2. Move the GSASIIscriptable.py to another directory within the GSAS library such as directly under the parent folder. the rough location of GSASIIscriptable.py from the parent directory of the library will be: /gsas2/GSAS-II/GSASII/GSASIIscriptable.py for newer versions of GSAS-II
- [x] On the GSAS II tab in the Engineering Diffraction interface, click refine after following step 3. You should receive an import error that specified that GSASIIscriptable.pyhas been found, but an import error occurred similar to:
```
 File "/mantidproject/mantid/qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/call_G2sc.py", line 200, in find_gsasii
   raise ImportError(
ImportError: GSASIIscriptable (GSASIIscriptable.py) module found in /gsas2 but could not be imported
```
- [x] Move the GSASIIscriptable.py back to it's original location: ~`/gsas2/GSAS-II/GSASII/GSASIIscriptable.py`


#### Older Versions of GSASII (versions older than `5758`. Version `4579` is installed on IDAaaS for EnginX)
1. Login to [IDAaaS](https://isis.analysis.stfc.ac.uk/) and create a new EnginX workspace.
2. Check the version of GSAS in the workspace is `4579`. This can be checked before launching the workspace by clicking on the setting cog at the top right of the workspace under the "Included Software" subheading.
3. Launch the EnginX workspace.
4. Open a terminal `Applications >> System >> Terminal`
5. Execute the following command: 
```bash
mamba create -n mantid-test -c mantid/label/new_gsas_path_search_test -c mantid/label/nightly mantidworkbench
``` 
`-c mantid/label/nightly` is required due to a an mslice version being pinned which will prevent you from creating a valid environment otherwise.
6. Activate your environment `mamba activate mantid-test` and open workbench `workbench` from the CLI.
7. Repeat tests performed for newer versions of GSAS ("_Newer Versions of GSASII (`5758` and later)_").

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

JackEAllen committed:
Refactor Gsas2InputParameters Structure to Collate Similar Attributes

Refactor Gsas2InputParameters to collate similar attrbiutes, update GSAS2Model and call_G2sc.py references to new GSAS2Model structure

JackEAllen committed:
Create DataClasses and GSAS2Handler Class

Create Dataclasses to group releate properties together, reduce the need for boilerplate code, provide clear structure and begin to add additional type safety. Create initial structure of GSAS2Handler class to centralise the responsibilities of data validation, path management, and general encapsulation of logic relating to GSAS-II input parameters and configuration in an effort to centralise logic related to loading of  GSAS-II.

JackEAllen committed:
Add Grouped Path Search methods to GSAS2Handler

Refactor existing GSAS2 paths searched for within GSAS2Model into GSAS2Handler class. Add ability to search for GSAS2Scriptable as a property in GSAS2Handler to make search for GSAS2Scriptable in call_G2sc.py redundant. This improves centralisation of GSAS2 path handling

JackEAllen committed:
Condense GSAS2Model init() and clear_input_componenets() by making use of dataclasses

JackEAllen committed:
Refactor implementation of subproccess call preperation for GSASII

improve readability, reusability, error handling and general maintainability of call to GSASII

JackEAllen committed:
Refactor to Use Provided gsasii_scriptable_path from input_dict

Create  method and remove  and  from  as part of refactor that add  to

JackEAllen committed:
Add Types

JackEAllen committed:
Move GSAS Handler and Related Data Classes out of Model

JackEAllen committed:
Types

JackEAllen committed:
Update Dataclasses and Types and General Cleanup

JackEAllen committed:
Update References to GSAS2Model Properties

JackEAllen committed:
Update Test References to GSAS2Model Properties

JackEAllen committed:
Create Tests for GSAS2Handler

Create Tests from GSAS2Handler including moving limited_rglob tests from test_call_G2sc.py into TestGSAS2Handler

JackEAllen committed:
Remove limited_rglob tests and add tests for import_gsasii

JackEAllen committed:
Test User Provided X Limits are Formatted by run_mode. Add XLimitsState Dataclass Mock to Model Mock

JackEAllen committed:
Release Notes

sf1919 commented on docs/source/release/v6.13.0/Diffraction/Engineering/New_features/39139.rst:
```diff
@@ -0,0 +1 @@
+- `#38882 <https://github.com/mantidproject/mantid/issues/39139>`_ : Refactor how ``GSAS-II`` files are found to avoid use of hard-coded paths which could cause issues going forward should other files neccessary for import of ``GSAS-II`` be relocated in the future like ``GSASIIscriptable.py`` was for versions 5758 and later.
```
The link is to 39139 but the text is 38882. Is that deliberate?

JackEAllen replied:
No, this is just copied over as a placeholder while the PR is still a draft. I will be updating this before stating it is ready for review.

sf1919 replied:
I'm just wondering if this is what is causing the docs build to fail.

JackEAllen replied:
Just pushed up an updated version so we can check. If it passes, I would be interested to know exactly how the docs failed. If it's because of a duplicate issue tag, I wonder if this is desirable functionality or not for the rare edge case where multiple PRs link to the same issue. This would could be a possibility if the PRs made sense to be broken into sections and have separate release notes to avoid having one really large PR and equally large release note to describe changes if all the work is related to one issue. - though there is a good argument that if this ever occurred, that separate sub-issues should be created.

RichardWaiteSTFC reviewed (changes requested):
Code looks great as far as I can tell! In particular the new `GSAS2Handler` makes it a lot see all the settings related to GSAS and now separates them into refinement and config classes rather than storing them all in a load of attributes of the model!

Just had a quick test on windows and I'm getting an error with no message when I click refine following Test 11 instructions
```
CreateSampleWorkspace started
CreateSampleWorkspace successful, Duration 0.01 seconds
LoadCIF started
LoadCIF successful, Duration 0.03 seconds
DeleteWorkspace started
DeleteWorkspace successful, Duration 0.00 seconds
GSAS-II call failed with error: 
```
I'm using a new-ish version of GSAS-II (2872d9 from 18-Feb-2025)
with the path to GSAS-II in the settings as 
`C:/Users/xhg73778/AppData/Local/gsas2full`
which works on main.
Can you reproduce this?

Sorry for posting this now but I won't be around Monday AM) - it's not urgent!

JackEAllen committed:
New method to prepare GSAS-II environment variables

Create method to correctly prepare environment variables for GSAS-II subprocess based on OS Platform

JackEAllen committed:
Test _prepare_gsas2_environment correctly sets environment variables based on platform

JackEAllen committed:
Refactor set_binaries to construct a more specific Windows-formatted list of binaries

JackEAllen committed:
Improve GSASII import error messages and try multiple  import strategies to improve cross OS compatability

JackEAllen committed:
Update Test for Windows Binaries

RichardWaiteSTFC commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/test/test_gsas_handler.py:
```diff
@@ -0,0 +1,322 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2024 ISIS Rutherford Appleton Laboratory UKRI,
```
I always forget to change the date here as well - it should be 2025 not 2024!

JackEAllen replied:
Good point! I'll update these now. I was unsure if we still updated these or not. In Mantid Imaging we have stopped updating them like other open source libraries such as numpy.

RichardWaiteSTFC commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/gsas2_handler.py:
```diff
@@ -0,0 +1,327 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2022 ISIS Rutherford Appleton Laboratory UKRI,
```
2025 I think!

RichardWaiteSTFC reviewed (dismissed):
Thanks for this - I have tested on windows and IDAaaS with multiple phases. When modifying the fit range in the instructions I did find some errors but they are also rpesent on main (see issue #39349).

Some really minor changes but it all looks really good to me!
Feel free to push back as near code freeze and I will approve!

JackEAllen committed:
Update copyright year

JackEAllen commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py:
```diff
@@ -173,88 +277,140 @@ def initial_validation(self, project, load_params):
+            config=config,
+        )
+
+        gsas2_handler.os_platform = platform.system()
```
If I remember correctly, this is functionality left over from the original implementation. Having looked at it a bit more, I agree with you. given the code freeze is tomorrow. I think I will leave it this time if you agree and move it as part of a further refactor I am planning going forward?  (I will still be contributing to diffraction when I move back to Mantid Imaging as there are a few things i would like to finish off

JackEAllen replied:
Had a change of mind overnight so this line has been dropped and now the the `os_platform` is set on initialisation of the `GSAS2Handler` class

JackEAllen committed:
Improve clarity of variables in docstring for XLimitsState dataclass

JackEAllen committed:
Set OS Platform on GSAS2Hander Initialisation

warunawickramasingha reviewed (dismissed):
Followed the test steps and results look good and the code changes also look fine and happy to approve!

warunawickramasingha reviewed (approved):
Re approving for the release-next branch.
